File: Juego/Invasiones/fuente/Eventos/Teclado.cs
using System;
using System.Collections.Generic;
using System.Text;
using Tao.Sdl;

namespace Invasiones.Eventos
{
    class Teclado
    {
        #region Declaraciones

		public const int INTERVALO_ENTRE_REPETICIONES = 15;

        public const int TECLA_ARR = Sdl.SDLK_UP;
        public const int TECLA_ABJ = Sdl.SDLK_DOWN;
        public const int TECLA_IZQ = Sdl.SDLK_LEFT;
        public const int TECLA_DER = Sdl.SDLK_RIGHT;

        public const int TECLA_A = Sdl.SDLK_a;
        public const int TECLA_B = Sdl.SDLK_b;
        public const int TECLA_C = Sdl.SDLK_c;
        public const int TECLA_D = Sdl.SDLK_d;
        public const int TECLA_E = Sdl.SDLK_e;
        public const int TECLA_F = Sdl.SDLK_f;
        public const int TECLA_G = Sdl.SDLK_g;
        public const int TECLA_H = Sdl.SDLK_h;
        public const int TECLA_I = Sdl.SDLK_i;
        public const int TECLA_J = Sdl.SDLK_j;
        public const int TECLA_K = Sdl.SDLK_k;
        public const int TECLA_L = Sdl.SDLK_l;
        public const int TECLA_M = Sdl.SDLK_m;
        public const int TECLA_N = Sdl.SDLK_n;
        public const int TECLA_O = Sdl.SDLK_o;
        public const int TECLA_P = Sdl.SDLK_p;
        public const int TECLA_Q = Sdl.SDLK_q;
        public const int TECLA_R = Sdl.SDLK_r;
        public const int TECLA_S = Sdl.SDLK_s;
        public const int TECLA_T = Sdl.SDLK_t;
        public const int TECLA_U = Sdl.SDLK_u;
        public const int TECLA_V = Sdl.SDLK_v;
        public const int TECLA_W = Sdl.SDLK_w;
        public const int TECLA_X = Sdl.SDLK_x;
        public const int TECLA_Y = Sdl.SDLK_y;
        public const int TECLA_Z = Sdl.SDLK_z;

		public const int TECLA_RSHIFT = Sdl.SDLK_RSHIFT;
		public const int TECLA_LSHIFT = Sdl.SDLK_LSHIFT;
		public const int TECLA_MAYUSCULA = Sdl.SDLK_CAPSLOCK;
		public const int TECLA_BACKSPACE = Sdl.SDLK_BACKSPACE;
		public const int TECLA_ENTER = Sdl.SDLK_RETURN;
		public const int TECLA_ESC = Sdl.SDLK_ESCAPE;

        /// <summary>
        /// Solo las teclas que estan siendo apretadas estan guardadas aqui.
        /// </summary>
        private List<int> m_teclasApretadas;

        /// <summary>
        /// La instancia del teclado
        /// </summary>
        private static Teclado s_instancia;
        #endregion

        #region Constructores
        /// <summary>
        /// Cosntructor.
        /// </summary>
        private Teclado()
        {
            m_teclasApretadas = new List<int>();
        }
        #endregion

        #region Properties
        /// <summary>
        /// Devuelve la instancia de la clase.
        /// </summary>
        public static Teclado Instancia
        {
            get
            {
                if (s_instancia == null)
                {
                    s_instancia = new Teclado();
                }
                return s_instancia;
            }
        }

        /// <summary>
        /// Devuelve una lista con las teclas apretadas.
        /// </summary>
        public List<int> TeclasApretadas
        {
            get
            {
                return m_teclasApretadas;
            }
        }
        #endregion
    }
}

File: macOS/Invasiones/Invasiones/Eventos/Teclado.swift
// Eventos/Teclado.swift
// Puerto de Teclado.cs — singleton que rastrea teclas presionadas.
// SDL keycodes reemplazados por los keyCode de NSEvent (Carbon virtual key codes).

import AppKit

class Teclado {

    // MARK: - Constantes de teclas (Carbon virtual key codes en macOS)
    static let INTERVALO_ENTRE_REPETICIONES = 15

    static let TECLA_ARR      = 126   // kVK_UpArrow
    static let TECLA_ABJ      = 125   // kVK_DownArrow
    static let TECLA_IZQ      = 123   // kVK_LeftArrow
    static let TECLA_DER      = 124   // kVK_RightArrow

    static let TECLA_A = 0;  static let TECLA_B = 11; static let TECLA_C = 8
    static let TECLA_D = 2;  static let TECLA_E = 14; static let TECLA_F = 3
    static let TECLA_G = 5;  static let TECLA_H = 4;  static let TECLA_I = 34
    static let TECLA_J = 38; static let TECLA_K = 40; static let TECLA_L = 37
    static let TECLA_M = 46; static let TECLA_N = 45; static let TECLA_O = 31
    static let TECLA_P = 35; static let TECLA_Q = 12; static let TECLA_R = 15
    static let TECLA_S = 1;  static let TECLA_T = 17; static let TECLA_U = 32
    static let TECLA_V = 9;  static let TECLA_W = 13; static let TECLA_X = 7
    static let TECLA_Y = 16; static let TECLA_Z = 6

    static let TECLA_RSHIFT    = 60   // kVK_RightShift
    static let TECLA_LSHIFT    = 56   // kVK_Shift
    static let TECLA_MAYUSCULA = 57   // kVK_CapsLock
    static let TECLA_BACKSPACE = 51   // kVK_Delete
    static let TECLA_ENTER     = 36   // kVK_Return
    static let TECLA_ESC       = 53   // kVK_Escape

    // MARK: - Singleton
    private static var s_instancia: Teclado?

    static var Instancia: Teclado {
        if s_instancia == nil { s_instancia = Teclado() }
        return s_instancia!
    }

    // MARK: - Declaraciones
    /// Teclas actualmente presionadas (keyCode de NSEvent).
    private(set) var TeclasApretadas: [Int] = []

    // MARK: - Constructor (privado — singleton)
    private init() {}

    // MARK: - Metodos (llamados desde GameScene)
    func presionarTecla(_ keyCode: Int) {
        if !TeclasApretadas.contains(keyCode) {
            TeclasApretadas.append(keyCode)
        }
    }

    func soltarTecla(_ keyCode: Int) {
        TeclasApretadas.removeAll { $0 == keyCode }
    }

    func limpiarTeclas() {
        TeclasApretadas.removeAll()
    }
}
